trinity/tools/async_process_runner.py

```diff
@@ -0,0 +1,60 @@
+import asyncio
+import os
+import signal
+from typing import (
+    AsyncIterable,
+    Awaitable,
+    Callable,
+    Iterable,
+)
+
+
+class AsyncProcessRunner():
+
+    @classmethod
+    async def create_and_run(cls, cmds: Iterable[str], timeout_sec: int=10) -> 'AsyncProcessRunner':
+        runner = cls()
+        await runner.run(cmds, timeout_sec)
+        return runner
+
+    async def run(self, cmds: Iterable[str], timeout_sec: int=10) -> None:
+        proc = await asyncio.create_subprocess_exec(
+            *cmds,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            # We need this because Trinity spawns multiple processes and we need to take down
+            # the entire group of processes.
+            preexec_fn=os.setsid
+        )
+        self.proc = proc
+        asyncio.ensure_future(self.kill_after_timeout(timeout_sec))
+
+    async def iterate_stdout(self) -> AsyncIterable[str]:
+        async for line in self._iterate_until_empty(self.proc.stdout.readline):
+            yield line
+
+    async def iterate_stderr(self) -> AsyncIterable[str]:
+        async for line in self._iterate_until_empty(self.proc.stderr.readline):
+            yield line
+
+    async def _iterate_until_empty(
+            self,
+            awaitable_bytes_fn: Callable[[], Awaitable[bytes]]) -> AsyncIterable[str]:
+
+        while True:
+            line = await awaitable_bytes_fn()
+            # Leaving this print here allows running pytest with -s and observing the output
+            # There's probably a better way to handle that
+            print(line)
+            if line == b'':
+                return
+            else:
+                yield line.decode('utf-8')
+
+    async def kill_after_timeout(self, timeout_sec: int) -> None:
+        await asyncio.sleep(timeout_sec)
+        self.kill()
+        raise TimeoutError('Killed process after {} seconds'.format(timeout_sec))
+
+    def kill(self) -> None:
+        os.killpg(os.getpgid(self.proc.pid), signal.SIGKILL)
```

Comment:
This seems like it should use `trinity.utils.ipc.kill_process_id_gracefully`

Reply:
(with a little tweaking, in order to kill a whole process group, instead of a single process id)

Reply:
I now see that little `g` hanging around in there that I missed on first pass.

Reply:
Did you see my past comment here? https://github.com/ethereum/py-evm/pull/1049#discussion_r203996646